internal/jobobject/iocp.go

```diff
@@ -0,0 +1,83 @@
+package jobobject
+
+import (
+	"context"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct{}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(ctx context.Context, iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed to poll for job object notification")
+			continue
+		}
+		if ch, ok := jobMap[key]; ok {
+			// Check code and parse out relevant information related to that notification
+			// that we care about. For now all we handle is the message that all processes
+			// in the job have exited.
+			switch code {
+			case winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO:
+				// Send the AllProcessesExited notification. We don't remove the map entry here
+				// as for container scenarios for all intents and purposes this is the
+				// end of the containers life, but for possible future usecases for this package
+				// there's nothing stopping more processes from being added to the job and this
+				// being fired once the count reaches zero again. A call to JobObject.Close() will
+				// remove the entry as the handle is invalidated at this point and the job is
+				// finished for good.
+				ch <- AllProcessesExited{}
```

Comment:
What happens with the IOCP if this blocks (e.g. if no one is receiving from the channel)? Does the IOCP itself have some buffer size?

Reply:
I believe (from my escapade of debugging the multi job iocp issue) that each entry just gets added to a doubly linked list, so no explicit buffer size. It would just start filling up the queue https://docs.microsoft.com/en-us/windows/win32/api/ntdef/ns-ntdef-list_entry?redirectedfrom=MSDN

Reply:
Okay, so at least we presumably won't lose notifications. There is still a problem of if one channel is not being drained, then other channels will be blocked from getting notifications as well. We should think about how we could handle this. I wonder if there are any clues in existing Go projects (for instance, the event queue support in containerd).

Reply:
A couple of rough ideas:
- Spin up a goroutine to deliver each notification
  - More overhead
  - Could lead to out of order delivery if multiple notifications are waiting for a single job
- Send in a `select` block with a `default` clause
  - Less overhead
  - Would basically mean if nothing is listening, the notification is discarded

We could also combine either of these with allowing the client to specify a buffer size for the notification channel. That would give a bit more control of how many notifications should be cached, based on the client's receiving patterns.

Reply:
@jstarks do you have any thoughts on this? Basically we have a single IOCP handling notifications for all job objects. Each job object has its own notification channel. We are trying to figure out what the behavior should be to send a notification for a job object A, if nothing is receiving on its channel, so that we don't block job object B from also receiving notifications.

Reply:
I was going to say the buffer we have now on the channel gives us a little leeway here but the problem isn't solved at all with this obviously. I'm not following on the second proposal though, can you elaborate/share a code snippet? If a client isn't utilizing the channel then I feel they probably don't care about the notifications to begin with so the thought of discarding messages doesn't strike me as a no-go, although definitely not ideal. We care about them for the container case and we should always have a reader for the channel for this scenario but for other users of the package, I think a combination of a buffered channel and somehow dropping anything afterwards would be good.

Reply:
If you do a:

```
select {
    case ch <- v:
    default:
}
```

This implements a "non-blocking channel send". `v` will be sent to `ch` if there is buffer available or an active receiver. Otherwise, it will skip the send.

Reply:
I've never used that pattern, that.... seems like it might work fine for this actually. Do you have any strong thoughts about going with this? Like I said I feel like if the client doesn't utilize the channel then they likely don't care about the notifications to begin with. The buffer we have now on the message channel actually might serve a small purpose now, as it would at least hold a small amount of messages (possibly all of the ones that are sent in the case of a pretty well behaved job) so even if at a much later time they decide to drain the channel they'd at least have the first ten notifications.